src/socket.cpp

```diff
@@ -39,27 +56,72 @@ void CSocket::Init ( const quint16 iPortNumber, const quint16 iQosNumber, const
     WSAStartup ( MAKEWORD ( 1, 0 ), &wsa );
 #endif
 
-    // create the UDP socket
-    UdpSocket = socket ( AF_INET, SOCK_DGRAM, 0 );
-    //
-    const char tos = (char) iQosNumber; // Quality of Service
-    setsockopt ( UdpSocket, IPPROTO_IP, IP_TOS, &tos, sizeof ( tos ) );
+    memset ( &UdpSocketAddr, 0, sizeof ( UdpSocketAddr ) );
 
-    // allocate memory for network receive and send buffer in samples
-    vecbyRecBuf.Init ( MAX_SIZE_BYTES_NETW_BUF );
-
-    // preinitialize socket in address (only the port number is missing)
-    sockaddr_in UdpSocketInAddr;
-    UdpSocketInAddr.sin_family = AF_INET;
-    if ( strServerBindIP.isEmpty() )
+    if ( bEnableIPv6 )
     {
-        UdpSocketInAddr.sin_addr.s_addr = INADDR_ANY;
+        // try to create a IPv6 UDP socket
+        UdpSocket = socket ( AF_INET6, SOCK_DGRAM, 0 );
+        if ( UdpSocket == -1 )
+        {
+            // IPv6 requested but not available, throw error
+            throw CGenErr ( "IPv6 requested but not available on this system.", "Network Error" );
+        }
+
+        // The IPV6_V6ONLY socket option must be false in order for the socket to listen on both protocols.
+        // On Linux it's false by default on most (all?) distros, but on Windows it is true by default
+        const uint8_t no = 0;
+        setsockopt ( UdpSocket, IPPROTO_IPV6, IPV6_V6ONLY, (const char*) &no, sizeof ( no ) );
+
+        // set the QoS
+        const char tos = (char) iQosNumber; // Quality of Service
+        setsockopt ( UdpSocket, IPPROTO_IPV6, IPV6_TCLASS, &tos, sizeof ( tos ) );
+
+        UdpSocketAddr.sa6.sin6_family = AF_INET6;
+        UdpSocketAddr.sa6.sin6_addr   = in6addr_any;
+        UdpSocketAddrLen              = sizeof ( UdpSocketAddr.sa6 );
+
+        UdpPort = &UdpSocketAddr.sa6.sin6_port; // where to put the port number
+
+        // FIXME: If binding a dual-protocol interface to a specific address, does it cease to be dual-protocol?
```

Comment:
A socket listening on an IPv4 address is only listening for IPv4 traffic.  But the same _interface_ could have a separate socket listening on an IPv6 address, I believe.  The _interface_ has the MAC address (or equivalent).  The protocol has the IPv4 or IPv6 address that goes with the socket layer in use.

At least, that's my understanding.

Reply:
> A socket listening on an IPv4 address is only listening for IPv4 traffic. But the same _interface_ could have a separate socket listening on an IPv6 address, I believe.

Yes it could, but Jamulus only expects to listen on one socket. So I think that makes it unable to bind to a specific interface address and remain dual-stack. I don't think it's a problem, unless someone flags up a scenario where it is. The default of binding to `::` will allow it to receive both IPv4 and IPv6 traffic without needing two sockets.

Reply:
> Yes it could, but Jamulus only expects to listen on one socket.
> The default of binding to :: will allow it to receive both IPv4 and IPv6 traffic without needing two sockets.

How does an IPv4 client know what address to use?  Does the IPv6 socket "just know" it's got two addresses really?

Reply:
The IPv4 client will connect in the same way as at present, by being told the IP address, either explicitly, or from a hostname lookup, or by a directory server.

Binding an `AF_INET6` socket to `::` (declared as `in6addr_any`) is the IPv6 equivalent of binding an `AF_INET` socket to `0.0.0.0` (`INADDR_ANY`), which means it will accept any incoming packet that arrives on any interface. If the socket has been `setsockopt`ed to dual stack (as we do), it will accept both IPv4 and IPv6 packets.

An IPv4 address mapped to IPv6 uses `0000:0000:0000:0000:0000:ffff` as the first 96 bits, which denotes that the last 32 bits contain an IPv4 address in network byte order. Often expressed like `::ffff:192.168.1.1`. This mapping is used in `CSocket::SendPacket()` and checked for in `CSocket::OnDataReceived()`.